catboost/libs/metrics/metric.cpp

```diff
@@ -279,6 +279,39 @@ bool TMAPEMetric::IsMaxOptimal() const {
     return false;
 }
 
+/* SMAPE */
+
+TMetricHolder TSMAPEMetric::EvalSingleThread(
+    const TVector<TVector<double>>& approx,
+    const TVector<float>& target,
+    const TVector<float>& weight,
+    const TVector<TQueryInfo>& /*queriesInfo*/,
+    int begin,
+    int end
+) const {
+    CB_ENSURE(approx.size() == 1, "Metric SMAPE quantile supports only single-dimensional data");
+
+    const auto& approxVec = approx.front();
+    Y_ASSERT(approxVec.size() == target.size());
+
+    TMetricHolder error;
+    for (int k = begin; k < end; ++k) {
+        float w = weight.empty() ? 1 : weight[k];
+        error.Error += Abs(target[k] - approxVec[k]) / (Abs(approxVec[k]) + Abs(target[k]));
+        error.Weight += w;
+    }
+
+    return error;
+}
+
+TString TSMAPEMetric::GetDescription() const {
+    return ToString(ELossFunction::SMAPE);
+}
+
+bool TSMAPEMetric::IsMaxOptimal() const {
+    return true;
```

Comment:
should this be false?

Reply:
What I inferred from IsMaxOptimal is that the metric is bounded by finite values (Is my inference correct?). Since SMAPE is bounded on both ends I set this to true

Reply:
This means - if metric best possible value is as large as possible. This is used for overfitting detection, selecting best iteration with use_best_model and in visualization when drawing best point on graph.

Reply:
Oh, that means I interpreted incorrectly. Let me fix that.